packaging/hostname/hostname.1
.TH HOSTNAME 1 "2009-09-16" "net-tools" "Linux Programmer's Manual"

.SH NAME
hostname \- show or set the system's host name
.br
domainname \- show or set the system's NIS/YP domain name
.br
ypdomainname \- show or set the system's NIS/YP domain name
.br
nisdomainname \- show or set the system's NIS/YP domain name
.br
dnsdomainname \- show the system's DNS domain name
.br

.SH SYNOPSIS
.B hostname
.RB [ \-v ]
.RB [ \-a|\-\-alias ]
.RB [ \-d|\-\-domain ]
.RB [ \-f|\-\-fqdn|\-\-long ]
.RB [ \-A|\-\-all-fqdns ]
.RB [ \-i|\-\-ip-address ]
.RB [ \-I|\-\-all-ip-addresses ]
.RB [ \-s|\-\-short ]
.RB [ \-y|\-\-yp|\-\-nis ]
.br
.B hostname
.RB [ \-v ]
.RB [ \-b|\-\-boot ]
.RB [ \-F|\-\-file\ filename ]
.RB [ hostname ]
.br
.B hostname
.RB [ \-v ]
.RB [ \-h|\-\-help ]
.RB [ \-V|\-\-version ]
.PP
.B domainname
.RB [ nisdomain ]
.RB [ \-F\ file ]
.br
.B ypdomainname
.RB [ nisdomain ]
.RB [ \-F\ file ]
.br
.B nisdomainname
.RB [ nisdomain ]
.RB [ \-F\ file ]
.PP
.B dnsdomainname
.RB [ \-v ]

.SH DESCRIPTION
.B Hostname
is used to display the system's DNS name, and to display or set its hostname or
NIS domain name.

.SS "GET NAME"
When called without any arguments, the program displays the current
names:
.LP
.B hostname
will print the name of the system as returned by the
.BR gethostname (2)
function.
.LP
.B domainname
will print the NIS domainname of the system.
.BR domainname
uses the
.BR gethostname (2)
function, while
.BR ypdomainname
and
.BR nisdomainname
use the
.BR yp_get_default_domain (3).
.LP
.B dnsdomainname
will print the domain part of the FQDN (Fully Qualified Domain Name). The
complete FQDN of the system is returned with
.BR "hostname \-\-fqdn"
(but see the warnings in section
.B "THE FQDN"
below).

.SS "SET NAME"
When called with one argument or with the
.B \-\-file
option, the commands set the host name or the NIS/YP domain name.
.BR hostname
uses the
.BR sethostname (2)
function, while all of the three
.BR domainname,
.BR ypdomainname
and
.BR nisdomainname
use
.BR setdomainname (2).
Note, that this is effective only until the next reboot.
Edit /etc/hostname for permanent change.
.LP
Note, that only the super-user can change the names.
.LP
It is not possible to set the FQDN or the DNS domain name with the
.B dnsdomainname
command (see
.B "THE FQDN"
below).
.LP
The host name is usually set once at system startup in
.I /etc/init.d/hostname.sh
(normally by reading the contents of a file which contains
the host name, e.g.
.IR /etc/hostname ).

.SS THE FQDN
The FQDN (Fully Qualified Domain Name) of the system is the name that the
.BR resolver (3)
returns for the host name, such as,
.IR ursula.example.com .
It is usually the hostname followed by the DNS domain name (the part
after the first dot).  You can check the FQDN using
.B "hostname \-\-fqdn"
or the domain name using
.BR "dnsdomainname" .
.LP
You cannot change the FQDN with
.B hostname
or
.BR dnsdomainname .
.LP
The recommended method of setting the FQDN is to make the hostname be
an alias for the fully qualified name using
.IR /etc/hosts ,
DNS, or NIS. For example, if the hostname was "ursula", one might have a line in
.I /etc/hosts
which reads
.LP
.RS
127.0.1.1    ursula.example.com ursula
.RE
.LP
Technically: The FQDN is the name
.BR getaddrinfo (3)
returns for the host name returned by
.BR gethostname (2).
The DNS domain name is the part after the first dot.
.LP
Therefore it depends on the configuration of the resolver (usually in
.IR /etc/host.conf )
how you can change it. Usually the hosts file is parsed before DNS or
NIS, so it is most common to change the FQDN in
.IR /etc/hosts .
.LP
If a machine has multiple network interfaces/addresses or is used in a
mobile environment, then it may either have multiple FQDNs/domain names
or none at all. Therefore avoid using
.BR "hostname \-\-fqdn" ,
.BR "hostname \-\-domain"
and
.BR "dnsdomainname" .
.BR "hostname \-\-ip-address"
is subject to the same limitations so it should be avoided as well.

.SH OPTIONS
.TP
.I "\-a, \-\-alias"
Display the alias name of the host (if used). This option is deprecated
and should not be used anymore.
.TP
.I "\-A, \-\-all-fqdns"
Displays all FQDNs of the machine. This option enumerates all configured
network addresses on all configured network interfaces, and translates
them to DNS domain names. Addresses that cannot be translated (i.e. because
they do not have an appropriate reverse DNS entry) are skipped. Note that
different addresses may resolve to the same name, therefore the output may
contain duplicate entries. Do not make any assumptions about the order of the
output.
.TP
.I "\-b, \-\-boot"
Always set a hostname; this allows the file specified by \fI-F\fR to be
non-existant or empty, in which case the default hostname \fIlocalhost\fR
will be used if none is yet set.
.TP
.I "\-d, \-\-domain"
Display the name of the DNS domain.  Don't use the command
.B domainname
to get the DNS domain name because it will show the NIS domain name and
not the DNS domain name. Use
.B dnsdomainname
instead. See the warnings in section
.B "THE FQDN"
above, and avoid using this option.
.TP
.I "\-f, \-\-fqdn, \-\-long"
Display the FQDN (Fully Qualified Domain Name). A FQDN consists of a
short host name and the DNS domain name. Unless you are using bind or NIS
for host lookups you can change the FQDN and the DNS domain name (which is
part of the FQDN) in the \fI/etc/hosts\fR file. See the warnings in section
.B "THE FQDN"
above, and avoid using this option; use
.BR "hostname \-\-all-fqdns"
instead.
.TP
.I "\-F, \-\-file filename"
Read the host name from the specified file. Comments (lines starting with
a `#') are ignored.
.TP
.I "\-i, \-\-ip-address"
Display the network address(es) of the host name. Note that this works only
if the host name can be resolved. Avoid using this option; use
.BR "hostname \-\-all-ip-addresses"
instead.
.TP
.I "\-I, \-\-all-ip-addresses"
Display all network addresses of the host. This option enumerates all
configured addresses on all network interfaces. The loopback interface and IPv6
link-local addresses are omitted. Contrary to option \fI-i\fR, this option
does not depend on name resolution. Do not make any assumptions about the
order of the output.
.TP
.I "\-s, \-\-short"
Display the short host name. This is the host name cut at the first dot.
.TP
.I "\-v, \-\-verbose"
Be verbose and tell what's going on.
.TP
.I "\-V, \-\-version"
Print version information on standard output and exit successfully.
.TP
.I "\-y, \-\-yp, \-\-nis"
Display the NIS domain name. If a parameter is given (or
.B \-\-file name
) then root can also set a new NIS domain.
.TP
.I "\-h, \-\-help"
Print a usage message and exit.
.SH NOTES
The address families
.B hostname
tries when looking up the FQDN, aliases and network addresses of the
host are determined by the configuration of your resolver.
For instance, on GNU Libc systems, the resolver can be instructed to
try IPv6 lookups first by using the
.B inet6
option in
.BR /etc/resolv.conf .
.SH FILES
.B /etc/hostname
Historically this file was supposed to only contain the hostname and not the
full canonical FQDN. Nowadays most software is able to cope with a full FQDN
here. This file is read at boot time by the system initialization scripts to
set the hostname.
.LP
.B /etc/hosts
Usually, this is where one sets the domain name by aliasing the host name to
the FQDN.
.SH AUTHORS
Peter Tobias, <tobias@et-inf.fho-emden.de>
.br
Bernd Eckenfels, <net-tools@lina.inka.de> (NIS and manpage).
.br
Michael Meskes, <meskes@debian.org>
.br


vim/NEWS
vim (2:7.3.154+hg~74503f6ee649-1) unstable; urgency=low

  The vim-lesstif package has been removed in favor of the new vim-athena
  package.  The intent behind both packages is to provide a lighter-weight GUI
  package as well as one that allows using XFLD fonts.  The Athena toolkit,
  however,  has broader usage and reduces divergences with downstream
  distributions.

 -- James Vega <jamessan@debian.org>  Sun, 27 Feb 2011 12:45:40 -0500

vim (2:7.2c.000-1) experimental; urgency=low

  The autoindent option is no longer enabled by default because:
  - 'autoindent' is rather naïve and turning on filetype-specific indentation
    (via "filetype indent on") is usually a better choice.
  - not having any automatic indenting enabled by default prevents new users
    from encountering the bad indentation issue when pasting text.

  Filetype plugins are no longer enabled by default because certain actions
  (like preventing the loading of the default menus) *must* occur before
  filetype detection is enabled.  By enabling filetype plugins (and detection)
  in debian.vim, we are preventing the user from being able to make changes to
  Vim's behavior.  Filetype plugins can be re-enabled via
  "filetype plugin on".

 -- James Vega <jamessan@debian.org>  Thu, 07 Aug 2008 12:25:39 -0400

vim (1:7.1.285-1) unstable; urgency=low

  Due to popular demand (and general conformity with other filetype plugins),
  the debchangelog and debcontrol filetype plugins no longer enable folding by
  default.  If you wish to enable their folding, you can add the following
  commands to your ~/.vimrc for debchangelog and debcontrol, respectively:

    let g:debchangelog_fold_enable = 1
    let g:debcontrol_fold_enable = 1

 -- James Vega <jamessan@debian.org>  Sun, 30 Mar 2008 12:42:35 -0400

vim (1:7.1-022+1) unstable; urgency=low

  * /usr/share/vim/addons/ is no longer in the vim runtimepath

    This is intended to avoid arbitrarly clashes among addons installed
    system-wide and enabled by default, and to set up a common ground for Vim
    extensions of various kinds.  From now on vim addons won't be enabled by
    default upon installation but will need to be enabled on a per-addon basis
    either in a system-wide manner (affecting all users of the system) or by
    each user who wants to use them.  A tool exists for helping with addon
    management: it is called "vim-addons" and is shipped by the package
    "vim-addon-manager"; all vim addon packages should recommend it.

    For more information see the VIM packaging policy (available on the web at
    http://pkg-vim.alioth.debian.org/vim-policy.html/index.html) and the
    manual page of vim-addons, available in the vim-addon-manager package.

 -- Stefano Zacchiroli <zack@debian.org>  Thu, 19 Jul 2007 23:28:44 +0200


packaging/hostname/hostname.c
/*
 * hostname	This file contains an implementation of the command
 *		that maintains the host name and the domain name. It
 *		is also used to show the FQDN and the IP-Addresses.
 *
 * Usage:	hostname [-a|-A|-d|-f|-i|-I|-s|-y]
 *		hostname [-h|-V]
 *		hostname [-b] {name|-F file}
 *		dnsdomainname
 *		domainname
 *		ypdomainname
 *		nisdomainname
 *
 * Authors:	Peter Tobias <tobias@et-inf.fho-emden.de>
 *		Bernd Eckenfels
 *		Graham Wilson <graham@debian.org>
 *		Michael Meskes <meskes@debian.org>
 *
 *		This program is free software; you can redistribute it
 *		and/or  modify it under  the terms of  the GNU General
 *		Public  License as  published  by  the  Free  Software
 *		Foundation;  either  version 2 of the License, or  (at
 *		your option) any later version.
 *
 *		The localdomain and localhost functions are copyright
 *		(C) 1996 Free Software Foundation, Inc. and were
 *		written by Miles Bader <miles@gnu.ai.mit.edu> and
 *		Marcus Brinkmann <brinkmd@debian.org>.
 */

#include <sys/stat.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <ifaddrs.h>
#include <net/if.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <getopt.h>
#define __USE_GNU 1
#include <string.h>
#include <netdb.h>
#include <errno.h>
#include <ctype.h>
#include <err.h>
#include <rpcsvc/ypclnt.h>

#define VERSION "3.12"

enum type_t { DEFAULT, DNS, FQDN, SHORT, ALIAS, IP, NIS, NIS_DEF, ALL_FQDNS, ALL_IPS };

char *progname;

/*
 * Return the name of the nis default domain. This is just a wrapper for
 * yp_get_default_domain.  If something goes wrong, program exits.
 */
char *
localnisdomain()
{
	char *buf = 0;
	int myerror = 0;

	myerror = yp_get_default_domain(&buf);

	/* yp_get_default_domain failed, abort. */
	if (myerror) {
		printf("%s: %s\n", progname, yperr_string(myerror));
		exit (1);
	}

	return buf;
}

/*
 * Return the name of the localdomain. This is just a wrapper for
 * getdomainname, which takes care of allocating a big enough buffer, and
 * caches the result after the first call (so the result should be copied
 * before modification). If something goes wrong, program exits.
 */
char *
localdomain()
{
	char *buf = 0;
	size_t buf_len = 0;
	int myerror = 0;

	do {
		errno = 0;

		if (buf) {
			buf_len += buf_len;
			if ((buf = realloc (buf, buf_len)) == NULL)
				err(1, NULL);
		} else {
			buf_len = 128;        /* Initial guess */
			if ((buf = malloc (buf_len)) == NULL)
				err(1, NULL);
		}
	} while (((myerror = getdomainname(buf, buf_len)) == 0 && !memchr (buf, '\0', buf_len))
	        || errno == ENAMETOOLONG);

	/* getdomainname failed, abort. */
	if (myerror)
		err(1, NULL);

	return buf;
}

/*
 * Return the name of the localhost. This is just a wrapper for gethostname,
 * which takes care of allocating a big enough buffer, and caches the result
 * after the first call (so the result should be copied before modification).
 * If something goes wrong, program exits.
 */
char *
localhost()
{
	char *buf = 0;
	size_t buf_len = 0;
	int myerror = 0;

	do {
		errno = 0;

		if (buf) {
			buf_len += buf_len;
			if ((buf = realloc (buf, buf_len)) == NULL)
				err(1, NULL);
		} else {
			buf_len = 128;        /* Initial guess */
			if ((buf = malloc (buf_len)) == NULL)
				err(1, NULL);
		}
	} while (((myerror = gethostname(buf, buf_len)) == 0 && !memchr (buf, '\0', buf_len))
	        || errno == ENAMETOOLONG);

	/* gethostname failed, abort. */
	if (myerror)
		err(1, NULL);

	return buf;
}

void
usage(FILE *stream)
{
	fprintf(stream,
		"Usage: hostname [-v] [-b] {hostname|-F file}         set host name (from file)\n"
		"       hostname [-v] [-a|-A|-d|-f|-i|-I|-s|-y]       display formatted name\n"
		"       hostname [-v]                                 display host name\n"
		"\n"
		"       {yp,nis,}domainname [-v] {nisdomain|-F file}  set NIS domain name (from file)\n"
		"       {yp,nis,}domainname [-v]                      display NIS domain name\n"
		"\n"
		"       dnsdomainname [-v]                            display dns domain name\n"
		"\n"
		"       hostname -V|--version|-h|--help               print info and exit\n"
		"\n"
		"Program name:\n"
		"       {yp,nis,}domainname=hostname -y\n"
		"       dnsdomainname=hostname -d\n"
		"\n"
		"Program options:\n"
		"    -a, --alias            alias names\n"
		"    -A, --all-fqdns        all long host names (FQDNs)\n"
		"    -b, --boot             set default hostname if none available\n"
		"    -d, --domain           DNS domain name\n"
		"    -f, --fqdn, --long     long host name (FQDN)\n"
		"    -F, --file             read host name or NIS domain name from given file\n"
		"    -i, --ip-address       addresses for the host name\n"
		"    -I, --all-ip-addresses all addresses for the host\n"
		"    -s, --short            short host name\n"
		"    -y, --yp, --nis        NIS/YP domain name\n"
		"\n"
		"Description:\n"
		"   This command can get or set the host name or the NIS domain name. You can\n"
		"   also get the DNS domain or the FQDN (fully qualified domain name).\n"
		"   Unless you are using bind or NIS for host lookups you can change the\n"
		"   FQDN (Fully Qualified Domain Name) and the DNS domain name (which is\n"
		"   part of the FQDN) in the /etc/hosts file.\n");
	exit(-1);
}

/*
 * Check the format of a user-specified hostname. Uses the rules from RFC 1035,
 * section 2.3.1.
 */
int
check_name(char *name)
{
	int i, len = strlen(name);

	/* Handle leading and trailing hyphen now. */
	if (!len || !isalnum(name[0]) || !isalnum(name[len-1]))
		return 0;

	for (i = 0; i < len; i++) {
		if (!isalnum(name[i]) && name[i] != '-' && name[i] != '.')
			return 0;
		if (name[i] == '-' && (name[i - 1] == '.' || name[i + 1] == '.'))
			return 0;
		if (name[i] == '.' && name[i - 1] == '.')
			return 0;
	}

	return 1;
}

void
set_name(enum type_t type, char *name)
{
	int i;

	switch (type) {
	case DEFAULT:
		/* Whitespaces are invalid characters in a hostname. */
		/* Thus remove trailing and leading whitespaces. */
		while (isspace(*name)) { name++; }
		for (i = strlen(name) - 1; i >= 0 && isspace(name[i]); i--);       
		name[i+1] = '\0';

		/* Now check for a valid name. */
		if (!check_name(name))
			errx(1, "the specified hostname is invalid");

		if (sethostname(name, strlen(name))) {
			if (errno == EPERM)
				errx(1, "you must be root to change the host name");
			else if (errno == EINVAL)
				errx(1, "name too long");
		}
		break;

	case NIS:
	case NIS_DEF:
		if (setdomainname(name, strlen(name))) {
			if (errno == EPERM)
				errx(1, "you must be root to change the domain name");
			else if (errno == EINVAL)
				errx(1, "name too long");
		}
		break;

	default:
		/*
		 * Only the host name and the domain name can be set using this
		 * command.
		 */
		usage(stderr);
	}
}

void
show_name(enum type_t type)
{
	struct addrinfo *res;
	struct addrinfo hints;
	struct ifaddrs *ifa, *ifap;
	char *p;
	int ret;

	/* Handle a few cases specially. */
	switch(type)
	{
		case DEFAULT: 
			printf("%s\n", localhost());
			break;
		case SHORT:
			p = localhost();
			*(strchrnul(p, '.')) = '\0';
			printf("%s\n", p);
			break;
		case NIS:
			printf("%s\n", localdomain());
			break;
		case NIS_DEF:
			printf("%s\n", localnisdomain());
			break;
		case ALL_IPS:
		case ALL_FQDNS: {
			char buf[INET6_ADDRSTRLEN];
			int flags, ret, family, addrlen;

			/* What kind of information do we want from getnameinfo()? */
			flags = (type == ALL_IPS) ? NI_NUMERICHOST : NI_NAMEREQD;

			if (getifaddrs(&ifa) != 0)
				errx(1, "%s", strerror(errno));
			for (ifap = ifa; ifap != NULL; ifap = ifap->ifa_next) {
				/* Skip interfaces that have no configured addresses */
				if (ifap->ifa_addr == NULL)
					continue;
				/* Skip the loopback interface */
				if (ifap->ifa_flags & IFF_LOOPBACK)
					continue;
				/* Skip interfaces that are not UP */
				if (!(ifap->ifa_flags & IFF_UP))
					continue;

				/* Only handle IPv4 and IPv6 addresses */
				family = ifap->ifa_addr->sa_family;
				if (family != AF_INET && family != AF_INET6)
					continue;

				addrlen = (family == AF_INET) ? sizeof(struct sockaddr_in) :
								sizeof(struct sockaddr_in6);

				/* Skip IPv6 link-local addresses */
				if (family == AF_INET6) {
					struct sockaddr_in6 *sin6;

					sin6 = (struct sockaddr_in6 *)ifap->ifa_addr;
					if (IN6_IS_ADDR_LINKLOCAL(&sin6->sin6_addr) ||
							IN6_IS_ADDR_MC_LINKLOCAL(&sin6->sin6_addr))
						continue;
				}

				ret = getnameinfo(ifap->ifa_addr, addrlen,
						  buf, sizeof(buf), NULL, 0, flags);

				/* Just skip addresses that cannot be translated */
				if (ret != 0) {
				    if (type != ALL_FQDNS && ret != EAI_NONAME)
					errx(1, "%s", gai_strerror(ret));
				} else
					printf("%s ", buf);
			}
			printf("\n");
			freeifaddrs(ifa);
			break;
		}
		default:
			memset(&hints, 0, sizeof(struct addrinfo));
			hints.ai_socktype = SOCK_DGRAM;
			hints.ai_flags = AI_CANONNAME;

			p = localhost();
			if ((ret = getaddrinfo(p, NULL, &hints, &res)) != 0)
				errx(1, "%s", gai_strerror(ret));
			
			p = strchr(res->ai_canonname, '.');

			switch (type) {
			case ALIAS: {
				struct hostent *hp;
				int i;

				if ((hp = gethostbyname(localhost())) == NULL)
					errx(1, "%s", hstrerror(h_errno));

				for (i = 0; hp->h_aliases[i]; i++) {
					if (i > 0)
						printf(" ");
					printf("%s", hp->h_aliases[i]);
				}
				printf("\n");
				break;
			}

			case IP: {
				char buf[INET6_ADDRSTRLEN];
				int ret;

				struct addrinfo *walk;

				for (walk = res; walk != NULL; walk = walk->ai_next) {
					if ((ret = getnameinfo(walk->ai_addr, walk->ai_addrlen,
								buf, sizeof(buf), NULL, 0,
								NI_NUMERICHOST)) != 0)
						errx(1, "%s", gai_strerror(ret));

					if (walk != res)
						printf(" ");

					printf("%s", buf);
				}
				printf("\n");
				break;
			}

			case DNS:
				if (p != NULL)
					printf("%s\n", ++p);
				break;

			case FQDN:
				printf("%s\n", res->ai_canonname);
				break;

			default:
				break;
			}
			break;
	}
}

char *
read_file(char *filename, int boot)
{
	struct stat st;
  	FILE *fp;
	char *buf;

	if ((fp = fopen(filename, "r")) == NULL) {
		if (boot)
			return NULL;
		else
			err(1, NULL);
	}

	if (fstat(fileno(fp), &st) == -1
	   || (buf = (char *) malloc(st.st_size + 1)) == NULL)
		err(1, NULL);

	while (fgets(buf, st.st_size + 1, fp) != NULL) {
		char *p;

		if (buf[0] == '\n' || buf[0] == '#')
			continue;

		if ((p = strchr(buf, '\n')) != NULL)
			*p = '\0';
		break;
	}

	fclose(fp);
	return buf;
}

int
main(int argc, char **argv)
{
	char *file = NULL, *name = NULL;
	enum type_t type = DEFAULT;
	int boot = 0;
	int o;

	static const struct option long_options[] =
	{
		{"domain", no_argument, 0, 'd'},
		{"boot", no_argument, 0, 'b'},
		{"file", required_argument, 0, 'F'},
		{"fqdn", no_argument, 0, 'f'},
		{"all-fqdns", no_argument, 0, 'A'},
		{"help", no_argument, 0, 'h'},
		{"long", no_argument, 0, 'f'},
		{"short", no_argument, 0, 's'},
		{"version", no_argument, 0, 'V'},
		{"verbose", no_argument, 0, 'v'},
		{"alias", no_argument, 0, 'a'},
		{"ip-address", no_argument, 0, 'i'},
		{"all-ip-addresses", no_argument, 0, 'I'},
		{"nis", no_argument, 0, 'y'},
		{"yp", no_argument, 0, 'y'},
		{0, 0, 0, 0}
	};

	/* If called as 'dnsdomainname', by default show the DNS domain name. */
	progname = (rindex(argv[0], '/')) ? rindex(argv[0], '/') + 1 : argv[0];
	if (!strcmp(progname, "dnsdomainname"))
		type = DNS;
	else if (!strcmp(progname, "domainname"))
		type = NIS;
	else if (!strcmp(progname, "ypdomainname"))
		type = NIS_DEF;
	else if (!strcmp(progname, "nisdomainname"))
		type = NIS_DEF;

	while((o = getopt_long(argc, argv, "aAdfbF:h?iIsVvy", long_options, NULL)) != -1)
		switch (o) {
		case 'd':
			type = DNS;
			break;
		case 'a':
			type = ALIAS;
			break;
		case 'f':
			type = FQDN;
			break;
		case 'A':
			type = ALL_FQDNS;
			break;
		case 'i':
			type = IP;
			break;
		case 'I':
			type = ALL_IPS;
			break;
		case 's':
			type = SHORT;
			break;
		case 'y':
			type = NIS_DEF;
			break;
		case 'b':
			boot = 1;
			break;
		case 'F':
			file = optarg;
			break;
		case 'v':
			/* Does not do anything. */
			break;
		case 'V':
			printf("hostname %s\n", VERSION);
			return 0;
		case '?':
		case 'h':
			usage(stdout);
			break;
		default:
			usage(stderr);
		}

	/*
	 * Hostname may be read from a file, it's ok for this file to not
	 * exist or be empty if -b is given in which case we default to
	 * "localhost" if there is not one already set.
	 */
	if (file) {
		name = read_file(file, boot);
		if (boot && (name == NULL || name[0] == '\0')) {
			free(name);

			name = localhost();
			if (name[0] == '\0' || !strcmp(name,"(none)"))
				strcpy(name, "localhost");
		}
	}

	/* Otherwise the hostname is the first non-option argument. */
	if (optind < argc) {
		/*
		 * It is an error to specify a host name as an argument, and to
		 * be read from a file.
		 */
		if (name)
			usage(stderr);

		if ((name = strdup(argv[optind])) == NULL)
			err(1, NULL);
		optind++;
	}

	/* Check for any remaining arguments. */
	if (optind < argc)
		usage(stderr);

	if (name) {
		set_name(type, name);
		free(name);
	} else
		show_name(type);

	return 0;
}

/* vim: set noexpandtab: */
